### ISSUE TITLE
Write docs re: building blocks as small libraries

### ISSUE THREAD

**Issue by bassosimone** [measurement-kit/measurement-kit#1666]
# TL;DR

Since MK v0.9.0-alpha-10, I've started refactoring functionality into small, C++11, header only libraries that live into their own repository. The advantage of this strategy are (1) much faster development cycle (because CI converges much more quickly) and (2) clearly defined API boundaries and much less interdependence between moving parts as we have now.

All these libraries have a similar, blocking API designed to be easily useable from all the target languages that we care about (i.e. Go, ObjectiveC and Java). They are vendored inside MK headers and compiled as part of the MK build process.

I plan to refactor more and more functionality as such, and to progressively phase out existing MK code that is asynchronous (hence more difficult to deal with) and more complex to wrap. The fact that we don't need to pay the cost of asynchronous code has been acknowledged by all of us.

# On the complexity of wrapping

This is, for example, this excerpt from the API of the new GeoIP wrappers:

```C++
mkgeoip_settings_t *mkgeoip_settings_new(void);
void mkgeoip_settings_set_asn_db_path(mkgeoip_settings_t *settings, const char *path);
mkgeoip_results *mkgeoip_lookup(const mkgeoip_settings_t *settings);
const char *mkgeoip_results_get_probe_asn(const mkgeoip_results_t *results);
int64_t mkgeoip_results_get_logs_as_binary(
    const mkgeoip_results_t *results, const uint8_t **base, uint64_t *length);
```

In the above API, the most complex part to wrap is binary getting logs (that is why I have included it into the above excerpt of the API). Wrapping that function in ObjectiveC is as easy as:

```ObjectiveC
@implementation MkGeoipResults

- (NSData *)getLogs {
  const uint8_t *base = nullptr;
  uint64_t length = 0;
  if (!mkgeoip_results_get_logs_as_binary(self.resultsp, &base, &length)) return nil;
  if (length > NSUIntegerMax) return nil;
  return [NSData dataWithBytes:(const void *)base length:(NSUInteger)length];
}

@end  // implementation MkGeoipResults
```

(The reason why we treat logs as binary data is because they _may_ include binary data if headers sent back from the server when performing the real probe IP lookup included binary data; this should not happen but we've had crashes in the past - especially on Android - because we were attempting to stuff binary data into strings, which assume UTF8 both on Java and ObjectiveC, but not on Go, which tells you something on the quality of the design of Go.)

The corresponding JNI mocking is:

```C++
JNIEXPORT jbyteArray JNICALL
Java_mkgeoip_MkGeoipResults_GetLogs(JNIEnv *env, jclass, jlong handle) {
  if (env == nullptr || handle == 0) return nullptr;
  mkgeoip_results_t *results = (mkgeoip_results_t *)handle;
  const uint8_t *base = nullptr;
  uint64_t length = 0;
  if (!mkgeoip_results_get_logs_as_binary(results, &base, &length)) return nullptr;
  // Note: both Android and Java define jsize as jint
  if (length > INT_MAX) return nullptr;
  jbyteArray array = env->NewByteArray((jsize)length);
  if (array == nullptr) return nullptr;
  env->SetByteArrayRegion(array, 0, (jsize)length, (const jbyte *)base);
  return array;
}
```

Which is Java is:

```Java
public class MkGeoipResults {
  public byte[] getProbeIP() {
    return this.GetProbeIP(this.handle);
  }

  static final native byte[] GetProbeIP(long handle);
};
```

# The rough plan going forward

(This is subject to change and I cannot guarantee I'll keep the issue in sync) In MK v0.9.0-alpha.10 we will ship two vendored libraries, one wrapping CURL (`mkcurl.h`) and one wrapping libmaxminddb (`mkgeoip.h`). I will write wrappers for `mkgeoip.h` in Go, ObjectiveC, and Java/JNI in the specific repositories that we use to adapt MK to such languages. This is done to implement #1656.

After that, I will ty to follow this plan (the order is tentative):

1. finish off [measurement-kit/web-api-clients](https://github.com/measurement-kit/web-api-clients) by restructuring it using the above mentioned style and, specifically, adding support for fetching URL lists, so to close #1664. This library will probably be called `mkoonib.h` and will contain and expose the following functionality: deal with orchestra, talk with the bouncer, submit reports. This library will depend on `mkcurl.h` for performing any HTTP(s) operation;

2. restructure the test workflow (now in `s<Tab>/l<Tab>/engine.cpp`) to use `mkoonib.h` for any interaction with the OONI backend and leave the existing measurement engine untouched;

3. remove the large fraction of code in MK now made unnecessary by the previous change;

4. modify `mkcurl.h` to add support for "connect only" (where CURL does not send any request and gives you an untouched socket) such that we can reimplement both WebConnectivity and TCPConnect using `mkcurl.h` as backend. These two tests will be implemented inside of MK itself but they will be only using the `mkcurl.h` blocking API;

5. create small libraries for both getdns and c-ares and then use them to reimplement the tests that require setting up a custom DNS and/or processing of DNS messages, then go ahead and refactor the existing OONI tests that touch upon this functionality.

**Comment by bassosimone** [comment]
I am not going to provide super detailed updates regarding the above plan, because I believe that would be too much effort. Some details have changed since I wrote the text, but the gist of it is that we're moving forward with the plan and we will implement the spirit of the changes in here.

I will occasionally reference this issue with more fine grained tickets when they are related to this effort, such that it's possible to understand what is going on with this refactoring via links.

**Comment by bassosimone** [comment]
I would like to update this ticket to mention something that I learned by applying this approach:

1. the small library approach is definitely a good thing and increases the _code velocity_

2. the C API is actually a bit of overhead, and should actually only be used for pieces of code that we want to export to ObjectiveC/JNI/Go (e.g. orchestra)

3. what's more, it's probably suboptimal to have the C API and its C++ implementation in the same place, and it would likely be better to have a C++ header (as an internal bit of implementation) and then C wrappers around it. Keeping them separate also increases the _code velocity_

All the rest that I said above still holds, we just need to change slightly the way in which this strategy is executed. Also, as stated, I am not keeping up to date the rough plan described above, as this is probably too much overhead and can deduced from looking at the stream of PRs.

**Comment by bassosimone** [comment]
I believe all the knowledge in this ticket should be moved to proper MK documentation.

**Comment by bassosimone** [comment]
Xref: https://github.com/measurement-kit/mkcurl/pull/21

**Comment by bassosimone** [comment]
I am going to remove this issue from v0.10.0, because anything that could have been done in that release to move forward this issue has IMHO been done. Looking forward to do more in v0.11.0.

**Comment by bassosimone** [comment]
I've changed the title such that I'll not postpone this task further. It's just to write documentation, so this can actually be done quite easily. Actually this can even go in v0.11.0.

**Comment by bassosimone** [comment]
This is superseded by switching to Golang.
